File: pyjoulescope_driver/entry_points/statistics.py
# Copyright 2022 Jetperch LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from pyjoulescope_driver import Driver
import time


def parser_config(p):
    """Display Joulescope measurement statistics."""
    p.add_argument('--verbose', '-v',
                   action='store_true',
                   help='Display verbose information.')
    p.add_argument('device_path',
                   nargs='*',
                   help='The target device for this command.')
    return on_cmd


def _on_statistics_value(topic, value):
    device = topic.split('/s/')[0]
    sample_id = value['time']['samples']['value'][0]
    i_avg, i_std, i_min, i_max = [value['signals']['current'][x]['value'] for x in ['avg', 'std', 'min', 'max']]
    v_avg, v_std, v_min, v_max = [value['signals']['voltage'][x]['value'] for x in ['avg', 'std', 'min', 'max']]
    p_avg, p_std, p_min, p_max = [value['signals']['power'][x]['value'] for x in ['avg', 'std', 'min', 'max']]
    charge = value['accumulators']['charge']['value']
    energy = value['accumulators']['energy']['value']
    print(f'{device},{sample_id},' +
          f'{i_avg},{i_std},{i_min},{i_max},' +
          f'{v_avg},{v_std},{v_min},{v_max},' +
          f'{p_avg},{p_std},{p_min},{p_max},' +
          f'{charge},{energy}')


def on_cmd(args):
    with Driver() as d:
        d.log_level = args.jsdrv_log_level
        devices = d.device_paths()
        if len(args.device_path):
            for device in args.device_path:
                if device not in devices:
                    print(f'device {device} not found')
                    return 1
            devices = args.device_path
        # display the column header
        print("#device,sampled_id," +
            "i_avg,i_std,i_min,i_max," +
            "v_avg,v_std,v_min,v_max," +
            "p_avg,p_std,p_min,p_max," +
            "charge,energy")
        for device in devices:
            d.open(device)
            d.publish(device + '/s/i/range/mode', 'auto')
            d.publish(device + '/s/stats/ctrl', 1)
            d.subscribe(device + '/s/stats/value', 'pub', _on_statistics_value)
        try:
            while True:
                time.sleep(0.025)
        except KeyboardInterrupt:
            pass

File: node_api/example/statistics.js
/*
 * Copyright 2024 Jetperch LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

const JoulescopeDriver = require("joulescope_driver");

function initialize(){
    const drv = new JoulescopeDriver();
    const device_paths = drv.device_paths();
    console.log('device_paths: ' + device_paths);
    if (0 == device_paths.length) {
        drv.finalize()
        return null;
    }
    let unsubs = [];

    device_paths.forEach((device_path) => {
        drv.open(device_path);

        var stats_cbk = (topic, value) => {
            topic_parts = topic.split('/');
            console.log(topic_parts[1] + '-' + topic_parts[2]
                + ',' + value['signals']['current']['avg']['value']
                + ',' + value['signals']['voltage']['avg']['value']
                + ',' + value['signals']['power']['avg']['value']
                + ',' + value['accumulators']['energy']['value']);
        }

        if (device_path.includes("/js110/")) {
            drv.publish(device_path.concat("/s/i/range/select"), "auto");
            unsubs.push(drv.subscribe(device_path.concat("/s/sstats/value"), 2, stats_cbk));
        } else {
            drv.publish(device_path.concat("/s/i/range/mode"), "auto");
            drv.publish(device_path.concat("/s/v/range/mode"), "auto");
            unsubs.push(drv.subscribe(device_path.concat("/s/stats/value"), 2, stats_cbk));
            drv.publish(device_path.concat('/s/stats/ctrl'), 1, 0);
        }
    });

    return () => {
        unsubs.forEach((unsub) => unsub());
        device_paths.forEach((device_path) => {
            drv.publish(device_path.concat('/s/stats/ctrl'), 0, 0);
            drv.close(device_path);
        })
        drv.finalize();
    }
}

finalize_cbk = initialize();
process.on('SIGINT', finalize_cbk);
